### ISSUE TITLE
view::group_by fails for views that mutate its state

### ISSUE THREAD

**Issue by gnzlbg** [ericniebler/range-v3#193]
Minimal example:

```c++
    auto rng = view::iota(0, 10) 
               | view::remove_if([](int n) { return n % 2 == 0; })
               | view::group_by([](int a, int b) { return a == b; /*illustrative*/});
    auto b = ranges::begin(rng);
```

This seems to be a general problem with `range_access` where the `range_access::begin_cursor(Rng & rng, int)` is not SFINAEd out but results in a hard error.

**Comment by gnzlbg** [comment]
I "worked around" this by constraining `view::group_by::begin_cursor() const` with the requirement that `const Rng` must model the `Range` concept as follows:

```c++
CONCEPT_REQUIRES(Callable<Fun const, range_common_reference_t<Rng>,
    range_common_reference_t<Rng>>() && Range<const Rng>())
cursor<true> begin_cursor() const 
{
    return {fun_, ranges::begin(rng_), ranges::end(rng_)};
}
```

since `const view::remove_if` does not model any range concept.

**Comment by ericniebler** [comment]
That sounds like the correct fix. Submit a pull request?

**Comment by gnzlbg** [comment]
Ah ok. I'll get to it then!

**Comment by gnzlbg** [comment]
closed by #196
